File: rust/advent-of-code-2021/src/solutions/five.rs
use regex::Regex;
use std::{
    fs::File,
    io::{BufRead, BufReader},
};

#[derive(Debug)]
pub struct Point {
    x: isize,
    y: isize,
}

#[derive(Debug)]
pub struct VentLine {
    from: Point,
    to: Point,
    is_diagonal: bool,
}

impl VentLine {
    pub fn new(from_x: isize, from_y: isize, to_x: isize, to_y: isize) -> VentLine {
        VentLine {
            from: Point {
                x: from_x,
                y: from_y,
            },
            to: { Point { x: to_x, y: to_y } },
            is_diagonal: from_x == to_x || from_y == to_y,
        }
    }
}

pub fn read_vents_from_file() -> Vec<VentLine> {
    let file = File::open("src/inputs/five.txt").expect("Cannot find/open input file for day 5");
    let reader = BufReader::new(file);

    let mut vent_lines: Vec<VentLine> = vec![];
    for line in reader.lines() {
        let current_line = line.expect("Cannot read line from input file");
        vent_lines.push(extract_vent_line(current_line));
    }
    return vent_lines;
}

fn extract_vent_line(s: String) -> VentLine {
    let vent_line_regex =
        Regex::new(r"([0-9]*),([0-9]*) -> ([0-9]*),([0-9*])").expect("Should be a valid regex");

    let (_, [from_x, from_y, to_x, to_y]) = vent_line_regex
        .captures(s.as_str())
        .map(|c| c.extract())
        .unwrap();
    return VentLine::new(
        from_x.parse::<isize>().unwrap(),
        from_y.parse::<isize>().unwrap(),
        to_x.parse::<isize>().unwrap(),
        to_y.parse::<isize>().unwrap(),
    );
}

File: golang/advent-of-code-2021/five/five.go
package five

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
)

func ReadVentFromFile() []VentLine {
	file, err := os.Open("five/five.txt")
	if err != nil {
		panic("Cannot find/open input file for day 5")
	}

	reader := bufio.NewReader(file)
	result := make([]VentLine, 0)
	for {
		currentLine, _, err := reader.ReadLine()
		if err != nil {
			break
		}
		ventRegex, err := regexp.Compile(`([0-9]*),([0-9]*) -> ([0-9]*),([0-9]*)`)
		if err != nil {
			panic("Invalid line extract regex")
		}
		result = append(result, extractVentLine(string(currentLine), ventRegex))
	}
	return result
}

func CountIntersections(ventLines *[]VentLine) int {
	intersections := 0
	pointMap := make(map[string]int)

	for _, ventLine := range *ventLines {
		if ventLine.IsDiagonal {
			continue
		}

		if ventLine.From.X == ventLine.To.X {
			for i := min(ventLine.From.Y, ventLine.To.Y); i <= max(ventLine.From.Y, ventLine.To.Y); i++ {
				key := fmt.Sprint(ventLine.From.X) + "," + fmt.Sprint(i)
				if _, ok := pointMap[key]; ok {
					pointMap[key] += 1
					continue
				}

				pointMap[key] = 1
			}
			continue
		}

		// For vertical vent line
		for i := min(ventLine.From.X, ventLine.To.X); i <= max(ventLine.From.X, ventLine.To.X); i++ {
			key := fmt.Sprint(i) + "," + fmt.Sprint(ventLine.From.Y)
			if _, ok := pointMap[key]; ok {
				pointMap[key] += 1
				continue
			}

			pointMap[key] = 1
		}
	}

	for _, value := range pointMap {
		if value >= 2 {
			intersections += 1
		}
	}

	return intersections
}

func extractVentLine(s string, regex *regexp.Regexp) VentLine {
	groups := regex.FindStringSubmatch(s)
	// ventLine := new(VentLine).Init(MustParseInt(groups[1]), MustParseInt(groups[2]), MustParseInt(groups[3]), MustParseInt(groups[4]))
	ventLine := NewVentLine(MustParseInt(groups[1]), MustParseInt(groups[2]), MustParseInt(groups[3]), MustParseInt(groups[4]))
	return *ventLine
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
